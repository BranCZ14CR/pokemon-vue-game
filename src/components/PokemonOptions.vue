<template>
  <div class="">
    <ul class="col-12 col-sm-12 col-md-10 col-lg-6 col-xl-6 mx-auto mb-5 mt-5">
      <li
        v-for="pokemon in pokemons"
        :key="pokemon.id"
        @click="$emit('selection', pokemon.id)"
        class="btn btn-warning"
      >
        {{ pokemon.name }}
      </li>
    </ul>
  </div>
</template>


<script>
export default {
  //Propiedades del componente
  props: {
    pokemons: {
      type: Array,
      required: true,
    },
  },
};
</script>


<style scoped>
/* Pokemon Options */
ul {
  list-style-type: none;
}

li {
  border-radius: 5px;
  cursor: pointer;
  margin: 5px 5px 5px 5px;
  width: 250px;
  color: white;
  text-transform: capitalize;
}
</style>