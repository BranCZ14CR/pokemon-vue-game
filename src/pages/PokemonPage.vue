<template class=" col-12 col-sm-12 col-md-10 col-lg-6 col-xl-6 mx-auto mb-5">
  <template v-if="!pokemon">
    <div class="text-center">
        <div class="spinner-border text-warning" role="status">
            <span class="sr-only">Loading...</span>
        </div>
    </div>
  </template>

  <div v-else>
    <h1 class="text-center mb-3 text-white">¿Quién es este pokémon?</h1>

    <PokemonPicture  :pokemon-id="pokemon.id" :show-pokemon="showPokemon" />

    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer($event)" />

    <template v-if="showAnswer">
      <h3 class="fade-in text-center mb-3 text-white">{{ message }}</h3>

      <button @click="newGame()" class="btn btn-info mb-3">Nuevo Juego</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  name: "PokemonPage",

  components: {
    PokemonPicture,
    PokemonOptions,
  },

  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
    };
  },

  methods: {
    //Obtener los nombres de los pokemos
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();

      //Obtener un pokemon de forma aleatoria
      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndInt];
    },

    //Comprobar la respuesta correcta
    checkAnswer(pokemonId) {
      this.showPokemon = true;
      this.showAnswer = true;

      if (pokemonId === this.pokemon.id) {
        this.message = `Felicitaciones atrapaste a un ${this.pokemon.name}`;
      } else {
        this.message = `Oh no se te ha escapado un ${this.pokemon.name}`;
      }
    },

    //Reiniciar el juego
    newGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemonArr = [];
      this.pokemon = null;
      this.mixPokemonArray();
    },
  },

  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style>
</style>