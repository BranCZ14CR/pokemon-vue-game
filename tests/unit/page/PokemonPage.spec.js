import PokemonOptions from "@/components/PokemonOptions"
import PokemonPicture from "@/components/PokemonPicture"
import PokemonPage from "@/pages/PokemonPage"
import { shallowMount, mount } from "@vue/test-utils"
import {pokemons} from '../mocks/pokemons.mock'

describe('Pokemon Page component', ()=>{

    let wrapper

    beforeEach(()=>{
        wrapper = shallowMount(PokemonPage)
    })

    test('Hacer match con Snapshot', ()=>{
        expect(wrapper.html()).toMatchSnapshot()
    })


    test('Debe llamar el mixPokemonArray al mounted', ()=>{
        const mixPokemonArray = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        shallowMount(PokemonPage)
        expect(mixPokemonArray).toHaveBeenCalled()
    })


    test('Hacer match con el snapshot cuando se cargan los pokemons', ()=>{
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                  pokemonArr: pokemons,
                  pokemon: pokemons[0],
                  showPokemon: false,
                  showAnswer: false,
                  message: "",
                };
            },
        })

        expect(wrapper.html()).toMatchSnapshot()
    })


    test('Mostrar los componentes de Pokemon Picture y Options', ()=>{

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                  pokemonArr: pokemons,
                  pokemon: pokemons[0],
                  showPokemon: false,
                  showAnswer: false,
                  message: "",
                };
            },
        })

        const pokemonP = wrapper.find('pokemon-picture-stub')
        const pokemonO = wrapper.find('pokemon-options-stub')

        expect(pokemonP).toBeTruthy()
        expect(pokemonO).toBeTruthy()
        
        expect(pokemonP.attributes('pokemonid')).toBe('1')
        expect(pokemonO.attributes('pokemons')).toBeTruthy()
    })

    test('Pruebas con checkAnswer', async ()=>{

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                  pokemonArr: pokemons,
                  pokemon: pokemons[0],
                  showPokemon: false,
                  showAnswer: false,
                  message: "",
                };
            },
        })

        await wrapper.vm.checkAnswer(1)

        expect(wrapper.find('h3').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h3').text()).toBe(`Felicitaciones atrapaste a un ${ pokemons[0].name }`)

        await wrapper.vm.checkAnswer(10)
        expect(wrapper.vm.message).toBe(`Oh no se te ha escapado un ${ pokemons[0].name }`)
    })

})