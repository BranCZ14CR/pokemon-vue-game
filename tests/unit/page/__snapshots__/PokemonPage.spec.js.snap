// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Page component Hacer match con Snapshot 1`] = `
<div class="text-center">
  <div class="spinner-border text-warning" role="status"><span class="sr-only">Loading...</span></div>
</div>
`;

exports[`Pokemon Page component Hacer match con el snapshot cuando se cargan los pokemons 1`] = `
<div>
  <h1 class="text-center mb-3 text-white">¿Quién es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
